src/resilience/backend/FenixBackend.cpp

```diff
@@ -0,0 +1,389 @@
+ * documentation and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the Corporation nor the names of the
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+ */
+#include "FenixBackend.hpp"
+
+#include <sstream>
+#include <fstream>
+#include <cassert>
+
+#include <fenix.h>
+
+#include <resilience/AutomaticCheckpoint.hpp>
+
+#ifdef KR_ENABLE_TRACING
+#include <resilience/util/Trace.hpp>
+#endif
+
+#include <fenix.h>
+
+// TODO: implement exception class
+//Kokkos::Impl::throw_runtime_exception( out.str() );
+#define FENIX_THROW( msg ) std::cerr << msg << std::endl
+
+namespace KokkosResilience
+{
+  class FenixClient
+  {
+    using serializer_t   = std::function< bool( std::ostream & ) >;
+    using deserializer_t = std::function< bool( std::istream & ) >;
+
+    struct ProtectedMemoryBlock
+    {
+      explicit ProtectedMemoryBlock( int id, const serializer_t &serializer, const deserializer_t &deserializer )
+          : m_id( id ), m_serializer( serializer ), m_deserializer( deserializer ) {}
+
+      int m_id;
+      serializer_t m_serializer;
+      deserializer_t m_deserializer;
+      std::vector<char> m_buffer;
+      std::size_t m_size;
+      bool m_protect = false;
+      bool m_registered = false;
+    };
+    
+  public:
+    // TODO Enable user configuration options, especially in regards to data group policy and checkpoint depth
+    FenixClient( MPI_Comm comm ) : m_comm( comm )
+    {
+      create_data_group();
+    }
+
+    ~FenixClient()
+    {
+      clear_registry();
+      delete_data_group();
+    }
+
+    void protect_member( int hash, serializer_t serializer, deserializer_t deserializer )
+    {
+      int member_id = g_current_member_id++;
+
+      auto item = m_registry.find( hash );
+      if ( item == m_registry.end() )
+      {
+        item = m_registry.emplace( std::piecewise_construct,
+                                   std::forward_as_tuple( hash ),
+                                   std::forward_as_tuple( member_id, serializer, deserializer ) ).first;
```

Comment:
```suggestion
        item = m_registry.try_emplace( hash, member_id, serializer, deserializer ).first;
```

This can be simplified a bit